packages/counterfactual_nft/contracts/CounterfactualNFT.sol

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+
+pragma solidity ^0.8.2;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC1155/ERC1155Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+import "./ICounterfactualNFT.sol";
+import "./AddressSet.sol";
+import "./external/IL2MintableNFT.sol";
+import "./external/IPFS.sol";
+import "./external/OwnableUpgradeable.sol";
+
+
+/**
+ * @title CounterfactualNFT
+ */
+contract CounterfactualNFT is ICounterfactualNFT, Initializable, ERC1155Upgradeable, OwnableUpgradeable, IL2MintableNFT, AddressSet
+{
+    event MintFromL2(
+        address owner,
+        uint256 id,
+        uint    amount,
+        address minter
+    );
+
+    bytes32 internal constant MINTERS = keccak256("__MINTERS__");
+    bytes32 internal constant DEPRECATED_MINTERS = keccak256("__DEPRECATED_MINTERS__");
+
+    address public immutable layer2Address;
+
+    modifier onlyFromLayer2
+    {
+        require(msg.sender == layer2Address, "not authorized");
+        _;
+    }
+
+    modifier onlyFromMinter
+    {
+        require(isMinter(msg.sender), "not authorized");
+        _;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(address _layer2Address)
+        initializer
+    {
+        layer2Address = _layer2Address;
+        // Disable implementation contract
+        _owner = 0x000000000000000000000000000000000000dEaD;
+    }
+
+    function initialize(address owner, string memory _uri)
+        public
+        override
+    {
+        require(_owner == address(0), "ALREADY_INITIALIZED");
+        _owner = owner;
+
+        if (bytes(_uri).length != 0) {
+            _setURI(_uri);
+        }
+    }
+
+    function setURI(string memory _uri)
+        public
+        onlyOwner
+    {
+        _setURI(_uri);
+    }
+
+    function mint(
```

Comment:
If baseURI is not empty, we should not allow people to mint from L1, and only one NFT will be allowed to mint on L2, 
because they share the same metadata uri